src/FSharp.Data.GraphQL.Server.AspNetCore/Giraffe/HttpHandlers.fs

```diff
@@ -1,155 +1,268 @@
-            if cancellationToken.IsCancellationRequested then
-                return (fun _ -> None) ctx
-            else
-                let options = ctx.RequestServices.GetRequiredService<GraphQLOptions<'Root>> ()
-                let executor = options.SchemaExecutor
-                let rootFactory = options.RootFactory
-                let serializerOptions = options.SerializerOptions
-                let deserializeGraphQLRequest () = task {
-                    try
-                        let! deserialized = JsonSerializer.DeserializeAsync<GraphQLRequest> (ctx.Request.Body, serializerOptions)
-                        return Ok deserialized
-                    with :? GraphQLException as ex ->
-                        logger.LogError (``exception`` = ex, message = "Error while deserializing request.")
-                        return Result.Error [ $"%s{ex.Message}\n%s{ex.ToString ()}" ]
-                }
 
-                let applyPlanExecutionResult (result : GQLExecutionResult) = task {
-                    let gqlResponse =
-                        match result.Content with
-                        | Direct (data, errs) -> GQLResponse.Direct (result.DocumentId, data, errs)
-                        | RequestError (problemDetailsList) -> GQLResponse.RequestError (result.DocumentId, problemDetailsList)
-                        | _ ->
-                            GQLResponse.RequestError (
-                                result.DocumentId,
-                                [
-                                    GQLProblemDetails.Create ("subscriptions are not supported here (use the websocket endpoint instead).")
-                                ]
+    let rec private moduleType = getModuleType <@ moduleType @>
+
+    let ofTaskIResult ctx (taskRes: Task<IResult>) : HttpFuncResult = task {
+        let! res = taskRes
+        do! res.ExecuteAsync(ctx)
+        return Some ctx
+    }
+
+    let ofTaskIResult2 ctx (taskRes: Task<Result<IResult, IResult>>) : HttpFuncResult =
+        taskRes
+        |> TaskResult.defaultWith id
+        |> ofTaskIResult ctx
+
+    /// Set CORS to allow external servers (React samples) to call this API
+    let setCorsHeaders : HttpHandler =
+        setHttpHeader "Access-Control-Allow-Origin" "*"
+        >=> setHttpHeader "Access-Control-Allow-Headers" "content-type"
+
+    let private handleGraphQL<'Root> (next : HttpFunc) (ctx : HttpContext) =
+        let sp = ctx.RequestServices
+
+        let logger = sp.CreateLogger moduleType
+
+        let options = sp.GetRequiredService<GraphQLOptions<'Root>>()
+
+        let toResponse { DocumentId = documentId; Content = content; Metadata = metadata } =
+
+            let serializeIdented value =
+                let jsonSerializerOptions = options.GetSerializerOptionsIdented()
+                JsonSerializer.Serialize(value, jsonSerializerOptions)
+
+            match content with
+            | RequestError errs ->
+                logger.LogInformation(
```

Comment:
I think loglevel `Information` is too verbose here. I can imagine having loglevel `Warning` here because some sort of error has occurred, but not an error on our side and instead an error on the client side.

Reply:
Ah, looks like it is a wrong version of the code. I used trace in the right one. I'll update tomorrow

Reply:
But it is not a warning, so warning is not suitable

Reply:
https://learn.microsoft.com/en-us/dotnet/core/extensions/logging?tabs=command-line#log-level

Reply:
fixed

Reply:
The description of _Warning_ on the page you linked above reads:

> _For abnormal or unexpected events. Typically includes errors or conditions that don't cause the app to fail._

Well, a request error is kind of unexpected. Normally, clients should make correct requests to the server. With GraphQL it's even more improbable to make malformed requests since the client GraphQL interface is often autogenerated by use of the introspection feature.
So, if there is a request error of sorts happening, as a server maintainer you'd probably be interested in being notified about that as something a little more "interesting" than an `Information` log entry. This way, it's easier to spot whether a client should be updated or if there is a malicious client somewhere trying to do some funny business.

Reply:
@ivelten @mickhansen @aggieben @njlr what do you think?